src/functions/connect.ts

```diff
@@ -0,0 +1,134 @@
+import { deviceCodes, messageConstants } from '../constants';
+import { buildRequest, request } from '../shared/functions';
+import { doesFetchWalletsOnLoad } from '../shared/predicates';
+import {
+  getPubKeyBytes,
+  getSharedSecret,
+  parseWallets,
+} from '../shared/utilities';
+import {
+  validateBaseUrl,
+  validateDeviceId,
+  validateKey,
+} from '../shared/validators';
+import { aes256_decrypt, getP256KeyPairFromPub } from '../util';
+
+export async function connect ({
+  id,
+  client,
+}: ConnectRequestFunctionParams): Promise<boolean> {
+  const { deviceId, key, baseUrl } = validateConnectRequest({
+    deviceId: id,
+    key: client.key,
+    baseUrl: client.baseUrl,
+  });
+  const url = `${baseUrl}/${deviceId}`;
+
+  const payload = await encodeConnectRequest(key);
+
+  const response = await requestConnect(payload, url);
+
+  const { isPaired, fwVersion, activeWallets, ephemeralPub } =
+    await decodeConnectResponse(response, key);
+
+  client.deviceId = deviceId;
+  client.ephemeralPub = ephemeralPub;
+  client.url = `${client.baseUrl}/${deviceId}`;
+  client.isPaired = isPaired;
+  client.fwVersion = fwVersion;
+  if (activeWallets) {
+    client.activeWallets = activeWallets;
+  }
+
+  // If we are paired and are on older firmware (<0.14.1), we need a follow up request to sync
+  // wallet state.
+  if (isPaired && !doesFetchWalletsOnLoad(client.getFwVersion())) {
+    await client.fetchActiveWallet();
+  }
+
+  return isPaired;
+}
+
+export const validateConnectRequest = ({
+  deviceId,
+  key,
+  baseUrl,
+}: ValidateConnectRequestParams): ValidatedConnectRequest => {
+  const validDeviceId = validateDeviceId(deviceId);
+  const validKey = validateKey(key);
+  const validBaseUrl = validateBaseUrl(baseUrl);
+  return {
+    deviceId: validDeviceId,
+    key: validKey,
+    baseUrl: validBaseUrl,
+  };
+};
+
+export const encodeConnectRequest = (key: KeyPair) => {
+  const deviceCode = deviceCodes.CONNECT;
+  const pubKeyBytes = getPubKeyBytes(key);
+  const payload = buildRequest(deviceCode, pubKeyBytes);
+  return payload;
+};
+
+export const requestConnect = async (
+  payload: Buffer,
+  url: string,
+): Promise<Buffer> => {
+  return request({ payload, url });
+};
+
+/**
+ * `decodeConnectResponse` will call `StartPairingMode` on the device, which gives the user 60 seconds to
```

Comment:
We should probably document all the major functions like this at some point

Reply:
Added #424